### PR TITLE
Support shift operators for variables

### LINKED ISSUE #33: Implement bit shifting
@hynese asks:
_Anyone know if it's possible to rotate a uint256 by 8 bits? I can't seem to find any suitable operators ;(_

(There is no lower level EVM opcode for SHL/SHR/ROR/ROL, I checked. So they must be emulated.)

It would be nice to implement:
https://github.com/ethereum/solidity/blob/develop/libsolidity/ExpressionCompiler.cpp#L1071
or this:
https://github.com/ethereum/solidity/blob/develop/libsolidity/ExpressionCompiler.cpp#L1067

Using this: (and testing, of course!)
_Shifting left by N is equivalent to multiplying by 2^N._
_Shifting right by N is (if you are using ones' complement) is the equivalent of dividing by 2^N and rounding to zero._
source: https://en.wikipedia.org/wiki/Bitwise_operation#Arithmetic_shift

SHL(input, bits) = input \* (2^bits)
SHR(input, bits) = input / (2^bits)  (then round to zero?)
Wikipedia has some code describing how to do rotate using shift, this is probably the most optimal way: https://en.wikipedia.org/wiki/Circular_shift#Implementing_circular_shifts
http://stackoverflow.com/questions/2835469/how-to-perform-rotate-shift-in-c#answer-2835503

(This also has applications for some cryptographic code, although such code is probably optimized by not using an emulated rotate/shift, it would simplify the logic of it.)

### PR DESCRIPTION
**Pull request by axic**
Add tests for:
- [x]    shifting constants
- [x]    smaller types
- [ ]    checks that high order bit cleanup is performed correctly
- [x]    right shift of negative left hand sides
- [x]    shifts by negative amounts (both for signed and unsigned right hand sides)
- [ ]    more complex usage of shifts in some actually useful snippet
- [x]    shift assignment (a <<= 2)

---

Do not merge yet - I would like to get feedback before finishing this work. Also some commits should be squashed later.

First two commits could be merged as those are only trivial fixes.

Regarding this rest:
- SHL seems solid
- SAR/SHR needs further testing (and not failing tests) and validation
- the tests need to be extended and perhaps it shouldn't be in the end-to-end test

(Fixes https://github.com/ethereum/solidity/issues/33.)

### CODE REVIEW

**Comment by chriseth** [comment]
Nice! Some comments:
- shifts are bit operations, they assume some representation of numbers in binary and thus it should not be possible to use them on integers
- signextend should be used as part of the "higher order bits cleanup" before the operation, i.e. appendShiftOperatorCode should not take a type argument.

**Comment by axic** [comment]
> shifts are bit operations, they assume some representation of numbers in binary and thus it should not be possible to use them on integers

You mean signed integers?

My understanding is the following: all shift operations should work on the internal representation of numbers, which is two's complement for `int`.

The different shifts:
1) SHL (shift left, <<) - Shouldn't care if the input is int/uint
2) SHR (shift right, >>>) - Shouldn't care if the input is int/uint,  `0` is added on the left
3) SAR (shift arithmetic right, >>) - Should care about sign, `1` is added on the left if it is signed (i.e. highest bit was 1, depends on the width of `intXX`)

The above is taken from Javascript and is what the Solidity parser had.  I am not fully convinced we need SAR - if omit it, I would suggest to have >> for SHR.

**Comment by VoR0220** [comment]
I would hold off on this until after the fixed point and rationals are implemented....there are some major changes that kind of dork what you're trying to do in your set up :)

**Comment by chriseth** [comment]
@axic no I mean integers in themselves. There is no such thing as "shifting an integer", it is just a number, you should not have to assume that it is in two's complement. At the point where you explicitly convert the integer into a `bytesNN`, it assumes two's complement representation, the notions of "left" and "right" get a meaning and it makes sense to "shift" in those directions.

Having said that: I realize how used people are to applying shift operators to integers...

So if we add shift operators, I would prefer to drop `>>>` and use `x >> y` to be identical to `x / 2**y`.

One advantage of that is that we can say that "shifting an integer" means multiplying / dividing by appropriate powers of two and thus get rid of the number representation issue.

Concerning the break with javascript: Solidity is typed, so the distinction between `>>` and  `>>>` is not really needed.

**Comment by VoR0220** [comment]
consider getting rid of the `>>>` token?

**Comment by axic** [comment]
@chriseth It is more clear now. Coming from a C background, I only have a notion of "integers" which behave like "bytesNN".

Do I understand it right that you want a left and right shift operator, denoted by << (SHL) and >> (SHR) respectively and that it behaves according to the type information. For signed integers the right shift would behave like an arithmetic right shift (= `x / 2**y`).

Is this correct?

I think @VoR0220 added support for shift operations on constants as a PR to my branch. I'll merge that soon.

**Comment by axic** [comment]
Following up with the previous question. Do we agree with the following?
- Have logical shifts only: `<<` (SHL) and `>>` (SHR) 
- Remove the `SHR` define
- Remove the `>>>` token

If yes, I'll push an update later today.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please don't move that up here, it is only three letters and the distance between declaration and usage is quite high.

Comment (VoR0220) [inline libsolidity/ast/Types.cpp]: my bad...I did that. Mainly because its being reused...figured it would be unnecessary to have multiple declarations.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: It should be an error if `other.m_value` is negative.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Also check negative here.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think it is better to pass only this flag to this function instead of the full type.
We might need two flags for the signedness of the first and the signedness of the second operator.

**Comment by axic** [comment]
As per our discussion, for the moment the following assumptions are made:
- `>>>` token is kept, but an exception is thrown in Types/ExpressionCompiler
- `SHL`, `SHR`, `SAR` defines are unchanged, but internally we'll only use `SHL` and `SAR` which denote `<<` and `>>` respectively.
- shifting with negative right value is prohibited (throws an error)

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I've changed it to a bool flag, but given every other method has the `Type` as a parameter, that seems to be more consistent.

**Comment by axic** [comment]
@chriseth 

Should I change this to include `Token::isShiftOp(c_op)`?

``` js
@@ -363,7 +363,7 @@ bool ExpressionCompiler::visit(BinaryOperation const& _binaryOperation)
                bool cleanupNeeded = false;
                if (Token::isCompareOp(c_op))
                        cleanupNeeded = true;
-               if (commonType.category() == Type::Category::Integer && (c_op == Token::Div || c_op == Token::Mod))
+               if (commonType.category() == Type::Category::Integer && (c_op == Token::Div || c_op == Token::Mod || Token::isShiftOp(c_op)))
                        cleanupNeeded = true;

                // for commutative operators, push the literal as late as possible to allow improved optimization
```

Also where do I add the check for negative right value? The same visitor? After cleanup? Something like this?

``` js
@@ -386,6 +386,11 @@ bool ExpressionCompiler::visit(BinaryOperation const& _binaryOperation)
                        leftExpression.accept(*this);
                        utils().convertType(*leftExpression.annotation().type, commonType, cleanupNeeded);
                }
+               if (Token::isShiftOp(c_op))
+               {
+                       IntegerType const& type = dynamic_cast<IntegerType const&>(*rightExpression.annotation().type);
+                       solAssert(!type.isSigned(), "Shift with negative right value is not supported.");
+               }
                if (Token::isCompareOp(c_op))
                        appendCompareOperatorCode(c_op, commonType);
                else
```

**Comment by chriseth** [comment]
I think it would be good to perform cleanup before every shift operation. The cleanup for the left operand is not necessary for left shifts, but I think it is fine to do it still.

The check that the right operand is not negative has to be done in asm. Even if the right operand has a signed type, it can still be positive.

Finally, I think we should ensure that the second operand has to be an integer (i.e. not a bytesN) - please add a test for that, not sure if we already do.

**Comment by chriseth** [comment]
I think it is good to promote the `appendShiftOperatorCode` function out of `appendOrdinaryBinaryOperatorCode` and call it at the same point where that is called because it requires some type information, at least for the second operand.

**Comment by VoR0220** [comment]
@axic what's the status of this?

**Comment by axic** [comment]
@VoR0220 I've been away, will look into my open Solidity issues this week.

**Comment by axic** [comment]
@chriseth 

Any idea how this could be broken?

```
/home/travis/build/ethereum/solidity/test/../test/libsolidity/SolidityExecutionFramework.h(71): fatal error in "bytes_index_access": Unknown exception thrown by m_compiler.compile(m_optimize, m_optimizeRuns)
```

Also, how to test with signed ints? There doesn't seem to be an `i256()`. Only `fromHex()` is the way?

Removed the catch-all and I get this:

```
libc++abi.dylib: terminating with uncaught exception of type std::bad_cast: std::bad_cast
```

**Comment by axic** [comment]
@chriseth:

> The check that the right operand is not negative has to be done in asm.

Do we check that in every single case or only if it a signed type?

I've added a piece of code to `appendShiftOperatorCode`, which invokes the errorTag.

**Comment by chriseth** [comment]
@axic signed integers: just use `u256(-1)`.
The test for negativeness has of course only be done for signed types.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:8676]: Please also add tests for:
- smaller types
- checks that high order bit cleanup is performed correctly
- right shift of negative left hand sides
- shifts by negative amounts (both for signed and unsigned right hand sides)
- more complex usage of shifts in some actually useful snippet
- shift assignment (`a <<= 2`)

**Comment by axic** [comment]
Add tests for:
- [x]    smaller types
- [ ]    checks that high order bit cleanup is performed correctly
- [x]    right shift of negative left hand sides
- [x]    shifts by negative amounts (both for signed and unsigned right hand sides)
- [ ]    more complex usage of shifts in some actually useful snippet
- [x]    shift assignment (a <<= 2)

File: docs/types.rst
Comment (chriseth) [inline docs/types.rst]: I think you should rebase this.

**Comment by axic** [comment]
I'll squash some of the commits and rebase against `develop` later.

**Comment by axic** [comment]
Putting this on hold until #1034 is merged (and this can be rebased).

**Comment by axic** [comment]
Rebased against #1034. Will merge the commits into a sane amount and then on to actually fixing this.

**Comment by axic** [comment]
@chriseth @pirapira I would appreciate a review of the test cases (don't mind the rest of the code, I'll clean that up).

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (pirapira) [inline test/libsolidity/SolidityEndToEndTest.cpp:8480]: Here `4266` is interpreted as a decimal, but you want it to be a hex.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (pirapira) [inline test/libsolidity/SolidityEndToEndTest.cpp:8497]: Similarly `-4266` is interpreted as a decimal, not hex.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp:8480]: Both the input and the output are decimal.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp:8497]: Both sides are decimal.

**Comment by chriseth** [comment]
By the way, vitalik came up with an interesting hack for shifting - we might want to check whether it is faster. It might have a negative impact on static analysis, though and does not work for signed shift:

`x >> y` can be implemented as

```
mstore(0, 0)
mstore(32, x)
mload(sub(32, y))
```

**Comment by axic** [comment]
Nice! It works in the context of Solidity, because the memory expansion cost was already paid for that area :)

It can also be applied to unsigned left shifts:

```
mstore(0, x)
mstore(32, 0)
mload(y)
```

The left shift takes 9 gas, the right shift takes 12, while `EXP` alone has a base fee of 10 and could become quite pricey after the next HF(s).

Cons:
- makes static analysis hard
- makes optimisations in JITs hard (i.e. evmjit)
- not sure about how to practically apply this (it only works if shifting by a multiple of 8)

Regarding the last point, we need:
a) to have an rvalue as a const so we can make the decision compile time
b) to add runtime check for the same

**Comment by axic** [comment]
This actually makes me think `MSTORE` and `MLOAD` are underpriced (in the case the area doesn't need to be extended) as they cost the same arithmetic instructions (add, sub, etc. - not taking exp under the same umbrella).

**Comment by axic** [comment]
I still need to squash some of the commits, but this **passes the tests now**.  Please review the test cases first :sparkles:

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Do we want to support shifting for addresses? I would say it is only valid for integer and bytes types.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:396]: Some comment as above about the mobile type also applies here.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1401]: It might be misleading to call these lvalue and rvalue.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:227]: Is it ok to just use `type` here? What about number constants that need to be converted to their mobile type first?

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Perhaps better to join these two lines?

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Perhaps better to join these two lines?

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Brace to the next line.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: In case this is a rational number, the check should be done at compile time (inside the type checker). We could add an assert here.

Also perhaps shorter: `dup2 0 slt` (or `0 dup2 slt` - please cehck)

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:369]: Strictly, we do not need to clean the left value in case of a left shift, but it is probably better to just clean everything.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:8503]: This test could be combined with one of the tests above. Also for other tests.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:8620]: Pleas also add tests for things like `1 << x` and `-1 << y`

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
Concerning the tests: It is probably not too important to have many regular shift tests because we assume that the VM is correct. What we could add in addition is higher order bits cleanup tests like

```
uint16 x = 0xffff;
x += 0x20;
x <<= 8;
x >>= 16;
```

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1401]: Why?

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1401]: Because lvalue and rvalue have special meaning in the context of compilers, but here, they only refer to the left and right side of the shift expression. It is true that in `a <<= b`, the `a` has to be an lvalue, but this is not true for `a << b`.

Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1401]: How would you name it then? It of course relates to the l/rvalue of the shift and not of the Solidity expression.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: It should be I think `0 dup3 slt`.

**Comment by axic** [comment]
Re tests: we would need better error reporting. Right now it just says which test failed, but wouldn't tell which assertion within. The reason I've separated them is to easily see what is failing.

**Comment by chriseth** [comment]
The test output should report the line number of the failure, so this should be sufficient unless you use loops.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp:1426]: Shall we also throw when the right hand side is bigger than the width of the type of the left hand side?

Another possibility is to throw for signed overflows.  This makes `(int3) 1 << 2` throw, because 4 is not representable in `int3`.

File: libsolidity/ast/Types.cpp
Comment (pirapira) [reply libsolidity/ast/Types.cpp]: We can choose to use the type of the left operand.  I don't see a reason why shifting a number should change the type of the number.  For me it's more natural if `uint8 << int32` stays to be `uint8` (why does a sign flag suddenly appear?).

**Comment by JMLX42** [comment]
I'm trying to add [Solidity support](https://github.com/promethe42/flatbuffers/blob/feature/solidity/solidity/flatbuffers.sol) for [FlatBuffers](https://google.github.io/flatbuffers/).
I'm stuck because bitshift operators are not implemented.

How close are we to have those operators working?

Thanks for the great work!

**Comment by ethernomad** [comment]
@promethe42 you can implement bitshifts very easily using exponential:

Left shift: `* 2**bits`
Right shift: `/ 2**bits`

**Comment by pirapira** [comment]
@axic I couldn't push my tests to this branch; will you cherry-pick https://github.com/pirapira/solidity/commit/ea5bca95b4e4716b9c3047ae8206c928076eb842 ?

I'm not sure about overflows caused by left shifts.  I wrote one test case to make sure they throw, but it's up to discussion.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (pirapira) [inline test/libsolidity/SolidityEndToEndTest.cpp]: I'm not sure about this one.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (pirapira) [inline test/libsolidity/SolidityEndToEndTest.cpp]: I'm even less sure about this one.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp]: I don't think it should throw on overflow as that would kind of defy a lot of the use cases.

Comment (VoR0220) [reply test/libsolidity/SolidityEndToEndTest.cpp]: me neither. Shift is one of those operators were you should only use it if you know what you are doing.

File: Changelog.md
Comment (chriseth) [inline Changelog.md:4]: Please rebase.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp:353]: As @pirapira suggested further below, this should not return the common type of the operands but rather the type of the left operand.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:214]: Please move that closer to the point where it is used.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1401]: Please don't use the words `rvalue` and `lvalue` because the `lvalue` might not be an lvalue.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think we need an "Unimplemented" assertion for fixed point types here.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1414]: Please add an assertion to check that `integerType()` is valid (and thus the rational number actually is an integer).

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1426]: I think the behaviour of `x << y` with regards to overflow should be identical to the behaviour of `x * 2**y`. I would argue that it should be identical in almost all other aspects, too.

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Is that comment accurate? The operator should be disabled at latest by the type checker.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1401]: Still waiting for a naming suggestion 😉

File: libsolidity/codegen/ExpressionCompiler.cpp
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Why would the type checker catch it? (It isn't catching it.)

File: libsolidity/ast/Types.cpp
Comment (axic) [inline libsolidity/ast/Types.cpp:353]: @chriseth Should we check here for the other side to be an address?

Comment (chriseth) [reply libsolidity/ast/Types.cpp:353]: I think you have to do this whole thing even before line 342. The following should be possible (please add a test):
```
uint x = 70;
int8 y = 2;
x << y;
```
`x` and `y` do not have a common type, and so currently, it would fail in line 345.

Yes, we should perform some more checks for the other side. I would say rational number is fine as long as it is an integer. Fixed point is fine, but we have to perform an integer check at run-time. Integer is fine as long as it is not an address.

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by pirapira** [comment]
I added a test case about cleaning up higher bits.

**Comment by chriseth** [comment]
Closing in Favour of #1487 (with a branch in this repository).

**Pull request closed without merging**
